fw/access_log.c

```diff
@@ -0,0 +1,430 @@
+	FIXED(" ")                                         \
+	UNTRUNCATABLE(version)                             \
+	FIXED("\" ")                                       \
+	UNTRUNCATABLE(status)                              \
+	FIXED(" ")                                         \
+	UNTRUNCATABLE(content_length)                      \
+	FIXED(" \"")                                       \
+	TRUNCATABLE(referer)                               \
+	FIXED("\" \"")                                     \
+	TRUNCATABLE(user_agent)                            \
+	FIXED("\"")
+
+static bool access_log_enabled = false;
+
+/* Use small buffer because printk won't display strings longer that ~1000 bytes */
+#define ACCESS_LOG_BUF_SIZE 960
+static DEFINE_PER_CPU_ALIGNED(char[ACCESS_LOG_BUF_SIZE], access_log_buf);
+
+/** Build string consists of chunks that belong to the header value.
+ * If header value is empty, then it returns "-" to be nginx-like.
+ *
+ * @param http_version is TFW_HTTP_VER_xx enum
+ * @param line is the request line from TfwHttpReq::h_tbl */
+static TfwStr
+get_http_header_value(char http_version, TfwStr *line)
+{
+	TfwStr result;
+	size_t len = line->len;
+	TfwStr *chunk = line->chunks, *end = chunk + line->nchunks;
+	static TfwStr empty_hdr = TFW_STR_STRING("-");
+	switch (http_version) {
+	case TFW_HTTP_VER_09:
+	case TFW_HTTP_VER_10:
+	case TFW_HTTP_VER_11:
+		/* Should never get plain string from parser */
+		if (TFW_STR_PLAIN(line))
+			return TFW_STR_EMPTY(line) ? empty_hdr : *line;
+		/* Skip over until ':' marker */
+		for (; chunk < end; chunk++) {
+			len -= chunk->len;
+			if (chunk->len == 1 && *chunk->data == ':')
+				break;
+		}
+		if (chunk == end)
+			return empty_hdr;
+		chunk++;
+		/* Skip possible whitespace blocks */
+		while (chunk < end && (chunk->flags & TFW_STR_OWS) != 0) {
+			len -= chunk->len;
+			chunk++;
+		}
+		break;
+	case TFW_HTTP_VER_20:
+		/* skip over chunks until HDR_VALUE one */
+		while (chunk < end && (chunk->flags & TFW_STR_HDR_VALUE) == 0) {
+			len -= chunk->len;
+			chunk++;
+		}
+		break;
+	default:
+		return empty_hdr;
+	}
+	if (chunk == end)
+		return empty_hdr;
+	TFW_STR_INIT(&result);
+	result.len = len;
+	result.chunks = chunk;
+	result.nchunks = end - chunk;
+	return result;
+}
+
+
+void
+do_access_log(TfwHttpResp *resp)
+{
+	do_access_log_req(resp->req, resp->status, resp->content_length);
+}
+
+#define HTTP_METHODS(XX) \
+	XX(COPY)         \
+	XX(DELETE)       \
+	XX(GET)          \
+	XX(HEAD)         \
+	XX(LOCK)         \
+	XX(MKCOL)        \
+	XX(MOVE)         \
+	XX(OPTIONS)      \
+	XX(PATCH)        \
+	XX(POST)         \
+	XX(PROPFIND)     \
+	XX(PROPPATCH)    \
+	XX(PUT)          \
+	XX(TRACE)        \
+	XX(UNLOCK)       \
+	XX(PURGE)
+
+static const struct {
+#	define MAX_HTTP_METHOD_NAME_LEN	10
+	char name[MAX_HTTP_METHOD_NAME_LEN];
+	u8 len;
+} http_methods[] = {
+#define XX(name) [TFW_HTTP_METH_ ## name] = { #name, sizeof(#name) - 1 },
+	HTTP_METHODS(XX)
```

Comment:
`HTTP_METHODS` is used only here, so there is not sense to define the macro.

`XX` is essentially a constant string initializer, so the code will be more clear with macro like `STR_INIT()` instead of `XX`.